### PR TITLE
Support Adding Application Egg to PySpark

### PR DESCRIPTION
**Pull request by rfrohock**
**Pull Request checklist**

- [x ] The commit(s) message(s) follows the contribution [guidelines](doc/contribution-guidelines.md#commit-message-format) ?
- [ ] Tests for the changes have been added (for bug fixes / features) ?
- [x] Docs have been added / updated (for bug fixes / features) ?

**Current behavior:** https://github.com/spark-jobserver/spark-jobserver/issues/940

**New behavior:** Support  Adding application egg to Pyspark context

**Other information**:
Didn't want to hardcode path version of examples egg, so did not include new unit test, will add if someone has a convention for adding a reference to the generated egg.  all of the existing tests do not make use of any egg files

---
This change is [Reviewable](https://reviewable.io/reviews/spark-jobserver/spark-jobserver/941)

### CODE REVIEW

File: job-server-python/src/python/sparkjobserver/subprocess.py [hunk 1]
```diff
@@ -98,6 +100,15 @@ def import_class(cls):
             exit_with_failure(
                     "Expected JavaSparkContext, SQLContext "
                     "or HiveContext but received %s" % repr(context_class), 2)
+
+    egg_path = os.environ.get("EGGPATH", None)
+    if egg_path is not None and sc is not None:
```
Comment (zeitos) [inline job-server-python/src/python/sparkjobserver/subprocess.py]: is empty a valid path?

Also: don't you want to fail if `sc is None`?

**Comment by zeitos** [comment]
@rfrohock Thanks! I think @noorul will need to review this PR before merging 'cos he did all the python stuff

**Comment by zeitos** [comment]
@rfrohock also: can you update the doc?

File: job-server-extras/src/main/scala/spark/jobserver/python/PythonJob.scala [hunk 2]
```diff
@@ -64,6 +64,7 @@ case class PythonJob[X <: PythonContextLike](eggPath: String,
         Process(
           Seq(sc.pythonExecutable, "-m", "sparkjobserver.subprocess", server.getListeningPort.toString),
           None,
+          "EGGPATH" -> eggPath,
```
Comment (noorul) [inline job-server-extras/src/main/scala/spark/jobserver/python/PythonJob.scala]: How are we going to upload the file to SJS node?

File: job-server-python/src/python/sparkjobserver/subprocess.py (hunk 1 above)
Comment (rfrohock) [reply job-server-python/src/python/sparkjobserver/subprocess.py]: will change the check for the path.  there are two cases JavaSparkContext and CustomContext that don't appear to set the SC, not sure why for the JavaSparkContext, in those cases the egg won't be added.  depending on the job you have submitted that might be ok.  all of the python tests worked without the egg being added before

File: job-server-extras/src/main/scala/spark/jobserver/python/PythonJob.scala (hunk 2 above)
Comment (rfrohock) [reply job-server-extras/src/main/scala/spark/jobserver/python/PythonJob.scala]: the application EGG file would have already been uploaded using standard job binary API.  this patch ensures the already loaded app path is set in the context of the shelled spark instance.

**Comment by noorul** [comment]
@rfrohock Can you update the doc?

**Comment by rfrohock** [comment]
@noorul Updated Python Doc, removed obsolete Context Factories, added new SessionContext.  Updated code examples.  Let me know if you need anything else

**Comment by noorul** [comment]
@rfrohock Did you forget to push the changes?

**Pull request merged**